repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/RequestHandler.scala

```diff
@@ -0,0 +1,524 @@
+    def buildTenantVector(defaultTenant: String, roleTenants: Seq[String], uriTenant: Option[String]): Vector[String] = {
+      val sendAllTenants = config.getTenantHandling.isSendAllTenantIds
+      val sendTenantIdQuality = Option(config.getTenantHandling.getSendTenantIdQuality)
+      val sendQuality = sendTenantIdQuality.isDefined
+      val defaultTenantQuality = sendTenantIdQuality.map(_.getDefaultTenantQuality).getOrElse(0.0)
+      val uriTenantQuality = sendTenantIdQuality.map(_.getUriTenantQuality).getOrElse(0.0)
+      val rolesTenantQuality = sendTenantIdQuality.map(_.getRolesTenantQuality).getOrElse(0.0)
+
+      var preferredTenant = defaultTenant
+      var preferredTenantQuality = defaultTenantQuality
+      uriTenant foreach { tenant =>
+        preferredTenant = tenant
+
+        preferredTenantQuality = if (defaultTenant.equals(tenant)) {
+          math.max(defaultTenantQuality, uriTenantQuality)
+        } else {
+          uriTenantQuality
+        }
+      }
+
+      if (sendAllTenants && sendQuality) {
+        val priorityTenants = uriTenant match {
+          case Some(tenant) => Vector(s"$defaultTenant;q=$defaultTenantQuality", s"$tenant;q=$uriTenantQuality")
+          case None => Vector(s"$defaultTenant;q=$defaultTenantQuality")
+        }
+        priorityTenants ++ roleTenants.map(tid => s"$tid;q=$rolesTenantQuality")
+      } else if (sendAllTenants && !sendQuality) {
+        Vector(defaultTenant) ++ roleTenants
+      } else if (!sendAllTenants && sendQuality) {
+        Vector(s"$preferredTenant;q=$preferredTenantQuality")
+      } else {
+        Vector(preferredTenant)
+      }
+    }
+
+    Option(config.getTenantHandling) map { tenantHandling =>
+      Option(tenantHandling.getValidateTenant) map { validateTenant =>
+        Option(validateTenant.getBypassValidationRoles) map {
+          _.getRole.asScala.intersect(validToken.roles).nonEmpty
+        } filter {
+          identity
+        } match {
+          case Some(_) => Success(None)
+          case None =>
+            expectedTenant match {
+              case Some(reqTenant) =>
+                val tokenTenants = Set(validToken.defaultTenantId) ++ validToken.tenantIds
+                tokenTenants.find(reqTenant.equals) match {
+                  case Some(uriTenant) => Success(Some(uriTenant))
+                  case None => Failure(InvalidTenantException("Tenant from URI does not match any of the tenants associated with the provided token"))
+                }
+              case None => Failure(UnparseableTenantException("Could not parse tenant from the URI"))
+            }
+        }
+      } match {
+        case Some(Failure(e)) => Failure(e)
+        case Some(Success(uriTenant)) => Success(buildTenantVector(validToken.defaultTenantId, validToken.tenantIds, uriTenant))
+        case None => Success(buildTenantVector(validToken.defaultTenantId, validToken.tenantIds, None))
```

Comment:
Yeah, I'm not sure I really like the nested return types. They are certainly expressive, but lead to rather dense and complex code blocks. Since David is gone, I'm going to blame him! :smiling_imp: 

I'll look for a way to simplify them, but I fear that may be a bit of a time cost, so I'll fix any outstanding issues before I do.